### PR TITLE
Improve Azure Pipelines Mono Repo Support

### LINKED ISSUE #22082: 🐛 Bug Report: `EntityAzurePipelinesContent` ignores `build-definition` when `project-repo` is specified
### 📜 Description

Consider an azure devops repo that is used as a mono repo for several applications, and you include the `project-repo` annotation to describe what repo the application code is in, but you also include the `build-definition` annotation because that repo contains multiple pipelines, only one of which applies to the current entity.

cc @awanlin 

### 👍 Expected behavior

The CI / CD dashboard should only show builds from the pipeline defined by the `dev.azure.com/build-definition` annotation 

### 👎 Actual Behavior with Screenshots

The pipeline dashboard shows pipeline runs from all pipelines associated with the repo

### 👟 Reproduction steps

Have an entity with the following annotations:
```
dev.azure.com/project-repo: MyOrg/MyRepo
dev.azure.com/project: MyOrg
dev.azure.com/build-definition: 1234
```

### 📃 Provide the context for the Bug.

It looks like the problem comes from https://github.com/backstage/backstage/blob/dbde3fa79fdc7bd57ee456c222b320ee39d871b0/plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts#L25 - if the `project-repo` annotation is present then the code quits before even checking the `build-definition` annotation. 

### 🖥️ Your Environment

yarn run v1.22.21
$ /home/alex/PlatformEngineering.Backstage/node_modules/.bin/backstage-cli info
OS:   Linux 5.15.133.1-microsoft-standard-WSL2 - linux/x64
node: v20.9.0
yarn: 1.22.21
cli:  0.24.0 (installed)
backstage:  1.20.3

Dependencies:
  @backstage/app-defaults                                          1.4.5
  @backstage/backend-app-api                                       0.5.8
  @backstage/backend-common                                        0.19.9
  @backstage/backend-dev-utils                                     0.1.2
  @backstage/backend-openapi-utils                                 0.1.0
  @backstage/backend-plugin-api                                    0.6.7
  @backstage/backend-tasks                                         0.5.12
  @backstage/catalog-client                                        1.4.6
  @backstage/catalog-model                                         1.4.3
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.2.0
  @backstage/cli                                                   0.24.0
  @backstage/config-loader                                         1.5.3
  @backstage/config                                                1.1.1
  @backstage/core-app-api                                          1.11.1
  @backstage/core-components                                       0.13.8
  @backstage/core-plugin-api                                       1.8.0
  @backstage/e2e-test-utils                                        0.1.0
  @backstage/errors                                                1.2.3
  @backstage/eslint-plugin                                         0.1.3
  @backstage/frontend-plugin-api                                   0.3.0
  @backstage/integration-aws-node                                  0.1.8
  @backstage/integration-react                                     1.1.21
  @backstage/integration                                           1.7.2
  @backstage/plugin-api-docs                                       0.10.1
  @backstage/plugin-app-backend                                    0.3.55
  @backstage/plugin-app-node                                       0.1.7
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.1
  @backstage/plugin-auth-backend-module-github-provider            0.1.4
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.4
  @backstage/plugin-auth-backend-module-google-provider            0.1.4
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.4
  @backstage/plugin-auth-backend                                   0.20.0
  @backstage/plugin-auth-node                                      0.4.1
  @backstage/plugin-azure-devops-backend                           0.4.4
  @backstage/plugin-azure-devops-common                            0.3.1
  @backstage/plugin-azure-devops                                   0.3.9
  @backstage/plugin-catalog-backend-module-azure                   0.1.26
  @backstage/plugin-catalog-backend-module-msgraph                 0.5.14
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.4
  @backstage/plugin-catalog-backend                                1.15.0
  @backstage/plugin-catalog-common                                 1.0.18
  @backstage/plugin-catalog-graph                                  0.3.1
  @backstage/plugin-catalog-node                                   1.5.0
  @backstage/plugin-catalog-react                                  1.9.1
  @backstage/plugin-catalog                                        1.15.1
  @backstage/plugin-entity-validation                              0.1.12
  @backstage/plugin-events-node                                    0.2.16
  @backstage/plugin-github-actions                                 0.6.8
  @backstage/plugin-home-react                                     0.1.5
  @backstage/plugin-home                                           0.5.11
  @backstage/plugin-org                                            0.6.17
  @backstage/plugin-permission-common                              0.7.10
  @backstage/plugin-permission-node                                0.7.18
  @backstage/plugin-permission-react                               0.4.17
  @backstage/plugin-proxy-backend                                  0.4.5
  @backstage/plugin-scaffolder-backend                             1.19.1
  @backstage/plugin-scaffolder-common                              1.4.3
  @backstage/plugin-scaffolder-node                                0.2.8
  @backstage/plugin-scaffolder-react                               1.6.1
  @backstage/plugin-scaffolder                                     1.16.1
  @backstage/plugin-search-backend-module-catalog                  0.1.11
  @backstage/plugin-search-backend-module-pg                       0.5.16
  @backstage/plugin-search-backend-module-techdocs                 0.1.11
  @backstage/plugin-search-backend-node                            1.2.11
  @backstage/plugin-search-backend                                 1.4.7
  @backstage/plugin-search-common                                  1.2.8
  @backstage/plugin-search-react                                   1.7.3
  @backstage/plugin-search                                         1.4.3
  @backstage/plugin-tech-radar                                     0.6.10
  @backstage/plugin-techdocs-backend                               1.9.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.2
  @backstage/plugin-techdocs-node                                  1.10.0
  @backstage/plugin-techdocs-react                                 1.1.13
  @backstage/plugin-techdocs                                       1.9.1
  @backstage/plugin-user-settings                                  0.7.13
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.4.5
  @backstage/theme                                                 0.4.4
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.7
Done in 0.55s.

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by R-Beck-2020**
## Hey, I just made a Pull Request!

Fixes #22082 
- When `build-definition`, `project` and `project-repo` annotations are provided, prefer the build-definition, as it is the most specific.
- `project` annotation is no longer required in the case `project-repo` is specified

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by afscrome** [comment]
Just thought about this some more, and I think we need to check whether `isAzurePipelinesAvailable` to account for this new combinations of annotations.  I think the logic works is right, but we should add the test cases to confirm.

https://github.com/backstage/backstage/blob/776ac8b283d4ad827baaaaa28136fa8fe10d6cd6/plugins/azure-devops/src/plugin.ts#L46

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for the contribution @R-Beck-2020, this looks good, left a couple small comments. One other thing that needs to be added is to update the `README` to mention how this works. 👍

**Comment by R-Beck-2020** [comment]
@awanlin - Cheers for the feedback, pushed the suggested changes

File: .changeset/wicked-elephants-scream.md
Comment (awanlin) [inline .changeset/wicked-elephants-scream.md:2]:
```suggestion
'@backstage/plugin-azure-devops': patch
```

Sorry, missed this before, nothing breaking in this PR.

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.test.ts
Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.test.ts:1]: @R-Beck-2020, this one might have been missed, not seeing a test?

File: plugins/azure-devops/README.md
Comment (awanlin) [inline plugins/azure-devops/README.md:80]: Can't seem to do a multi-line suggestion but can we use more explicit examples, say like this:

```yaml
dev.azure.com/project-repo: <project-with-source-code>/<my-repo>
dev.azure.com/build-definition: <build-definition-name>
dev.azure.com/project: <project-with-build-code>
```

Feel free to adjust if you have something better, I was just confused by the numbers and feel others might as well.

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks again, just a few more small comments

File: plugins/azure-devops/README.md
Comment (R-Beck-2020) [reply plugins/azure-devops/README.md:80]: RE: Missed test - Does this not cover it?
plugins/azure-devops/src/plugin.test.ts:72
`it('should be false when no annotations are present', () =>`

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.test.ts
Comment (R-Beck-2020) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.test.ts:1]: RE: Missed test - Does this not cover it?
plugins/azure-devops/src/plugin.test.ts:72
`it('should be false when no annotations are present', () =>`

Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.test.ts:1]: Right, that works when people are using `isAzurePipelinesAvailable` but what is they aren't? I know that we didn't use that when we only had Azure DevOps, once we started using GitHub then we did.

File: plugins/azure-devops/README.md
Comment (afscrome) [reply plugins/azure-devops/README.md:80]: @awanlin  For a multi line suggestion click on the top line, then shift click the second line before adding your comment.  Then your comment / suggestion should apply to multiple lines

image

Comment (R-Beck-2020) [reply plugins/azure-devops/README.md]: Changes pushed, hopefully this has everything we need

Comment (awanlin) [reply plugins/azure-devops/README.md]: Thanks @afscrome I did try that at the time but it just put the first line, probably user error though, lol

**Review by awanlin — APPROVED** [review]
Awesome! Many thanks again for this and following up my various comments @R-Beck-2020 🚀 

Now we just need a review from the maintainers

**Comment by awanlin** [comment]
@Rugvip @vinzscam this is ready for you to review 👍

**Comment by R-Beck-2020** [comment]
Merged latest changes into branch / hopefully all tests will pass now 🤞

**Review by Rugvip — APPROVED** [review]
Nice, thank you! 👍 🎉

**Pull request merged**
